File: app/src/main/java/com/thewhitewings/pouch/ui/MainViewModel.java
package com.thewhitewings.pouch.ui;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.thewhitewings.pouch.data.Note;
import com.thewhitewings.pouch.data.NotesRepository;
import com.thewhitewings.pouch.data.SortOption;
import com.thewhitewings.pouch.utils.Zone;

import java.util.List;

/**
 * ViewModel to interact with the {@link  NotesRepository}'s data source and the main activity.
 */
public class MainViewModel extends ViewModel {

    private static final String TAG = "MainViewModel";
    private final NotesRepository repository;
    private final LiveData<List<Note>> notesLiveData;
    private final MutableLiveData<Zone> currentZoneLiveData;
    private SortOption sortOption;
    private String searchQuery;

    /**
     * Constructor for {@link MainViewModel}.
     *
     * @param repository to be used for interacting with the data source
     */
    public MainViewModel(NotesRepository repository) {
        this.repository = repository;
        this.notesLiveData = repository.getAllNotes();
        this.currentZoneLiveData = new MutableLiveData<>(Zone.CREATIVE);
        this.sortOption = repository.getSortOption(Zone.CREATIVE);
        this.searchQuery = "";
    }

    /**
     * Get the notes live data.
     *
     * @return the notes live data
     */
    public LiveData<List<Note>> getNotesLiveData() {
        return notesLiveData;
    }

    /**
     * Get the current zone live data.
     *
     * @return the current zone live data
     */
    public LiveData<Zone> getCurrentZoneLiveData() {
        return currentZoneLiveData;
    }

    /**
     * Toggle the current zone.
     */
    public void toggleZone() {
        currentZoneLiveData.postValue(
                currentZoneLiveData.getValue() == Zone.CREATIVE ?
                        Zone.BOX_OF_MYSTERIES : Zone.CREATIVE
        );

        repository.toggleZone(currentZoneLiveData.getValue());
    }


    /**
     * Handle the selection of a sort option from the popup menu.
     *
     * @param menuItemId the id of the selected popup menu item
     */
    public void handleSortOptionSelection(int menuItemId) {
        SortOption selectedOption = SortOption.fromMenuItemId(menuItemId);
        if (selectedOption != null) {
            sortNotes(selectedOption);
        }
    }

    /**
     * Delete a note.
     *
     * @param note the note to be deleted
     */
    public void deleteNote(Note note) {
        repository.deleteNote(note);
    }

    /**
     * Search notes based on the given query and the current sort option.
     *
     * @param query the search query that represents a part of a note's title and/or body
     */
    public void searchNotes(String query) {
        searchQuery = query;
        repository.searchNotes(query, sortOption);
    }

    /**
     * Sort notes based on the given sort option.
     *
     * @param sortOption the sort option to be used for sorting the notes
     */
    public void sortNotes(SortOption sortOption) {
        this.sortOption = sortOption;
        repository.saveSortOption(sortOption, currentZoneLiveData.getValue());

        repository.sortNotes(sortOption, searchQuery);
    }


    /**
     * Factory class for creating instances of {@link MainViewModel}.
     */
    public static class MainViewModelFactory implements ViewModelProvider.Factory {
        private final NotesRepository repository;

        public MainViewModelFactory(NotesRepository repository) {
            this.repository = repository;
        }

        @NonNull
        @Override
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            if (modelClass.isAssignableFrom(MainViewModel.class)) {
                return (T) new MainViewModel(repository);
            }
            throw new IllegalArgumentException("Unknown ViewModel class");
        }
    }
}

File: app/src/main/java/com/thewhitewings/pouch/ui/MainViewModel.kt
package com.thewhitewings.pouch.ui

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProvider.AndroidViewModelFactory.Companion.APPLICATION_KEY
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import com.thewhitewings.pouch.PouchApplication
import com.thewhitewings.pouch.data.Note
import com.thewhitewings.pouch.data.NotesRepository
import com.thewhitewings.pouch.data.SortOption
import com.thewhitewings.pouch.data.getSortOptionFromId
import com.thewhitewings.pouch.utils.Zone
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

@OptIn(ExperimentalCoroutinesApi::class)
class MainViewModel(private val notesRepository: NotesRepository) : ViewModel() {

    private val _homeUiState = MutableStateFlow(HomeUiState())
    val homeUiState = _homeUiState.asStateFlow()

    init {
        viewModelScope.launch {
            notesRepository.getSortOptionFlow(_homeUiState.value.zone)
                .collect { sortOption ->
                    _homeUiState.update {
                        it.copy(sortOption = sortOption)
                    }
                }
        }

        viewModelScope.launch {
            combine(
                _homeUiState.map { it.sortOption },
                _homeUiState.map { it.searchQuery },
            ) { sortOption, searchQuery ->
                Pair(sortOption, searchQuery)
            }
                .flatMapLatest { (sortOption, searchQuery) ->
                    if (searchQuery.isEmpty()) {
                        notesRepository.getAllNotesStream(sortOption)
                    } else {
                        notesRepository.searchNotesStream(searchQuery, sortOption)
                    }
                }
                .collect { notesList ->
                    _homeUiState.update {
                        it.copy(notesList = notesList)
                    }
                }
        }
    }

    data class HomeUiState(
        val notesList: List<Note> = emptyList(),
        val zone: Zone = Zone.CREATIVE,
        val sortOption: SortOption = SortOption.NEWEST_FIRST,
        val searchQuery: String = ""
    )

    fun toggleZone() {
        notesRepository.toggleZone()
        viewModelScope.launch {
            val newZone =
                if (_homeUiState.value.zone == Zone.CREATIVE)
                    Zone.BOX_OF_MYSTERIES else Zone.CREATIVE

            _homeUiState.update {
                it.copy(
                    zone = newZone,
                    sortOption = notesRepository.getSortOptionFlow(newZone).first(),
                    searchQuery = ""
                )
            }
        }
    }

    fun updateSortOption(sortOptionId: Int) {
        val sortOption = getSortOptionFromId(sortOptionId)
        viewModelScope.launch {
            notesRepository.saveSortOption(sortOption, _homeUiState.value.zone)
        }
    }

    /**
     * Delete a note.
     *
     * @param note the note to be deleted
     */
    fun deleteNote(note: Note) {
        viewModelScope.launch {
            notesRepository.deleteNote(note)
        }
    }

    fun updateSearchQuery(newQuery: String) {
        _homeUiState.update {
            it.copy(searchQuery = newQuery)
        }
    }

    companion object {
        private const val TAG = "MainViewModel"

        /**
         * Factory for [MainViewModel] that takes [NotesRepository] as a dependency
         */
        val Factory: ViewModelProvider.Factory = viewModelFactory {
            initializer {
                val application = (this[APPLICATION_KEY] as PouchApplication)
                val notesRepository = application.notesRepository
                MainViewModel(notesRepository = notesRepository)
            }
        }
    }

}
